### PR TITLE
Add support for multitenancy

### LINKED ISSUE #2611: Multitenancy
What
===
Make CONSUL Multitenant

How
===
We suggest using the [apartment gem](https://github.com/influitive/apartment)

Here are a [couple](https://medium.freecodecamp.org/building-a-multi-tenant-app-is-easy-if-you-have-an-apartment-3465f6eda85b) of [articles](http://dev.edenspiekermann.com/2016/09/09/rails-multitenant-site/) and a [screencast](https://gorails.com/episodes/multitenancy-with-apartment) on how to set it all up

Basic things to take care of per tenant:
- [ ] Separate database
- [ ] Subdomain
- [ ] Settings
- [ ] Logo and styles
- [ ] Postal Codes
- [ ] Census

Screenshots
==

1_tenant_select

2_tenant_select

3_tenant_select

4_tenant_admin

### PR DESCRIPTION
**Pull request by javierm**
## References

* Closes #2611
* Continues pull request #2816

## Objectives

Make it possible to manage several tenants within one CONSUL installation

## Notes about current status

The basic support for multitenancy provided by this pull request is in the testing phase. See the future improvements section below for planned features which are **not** part of this pull request.

### Completed items:

* Update attachments code to use Active Storage instead of Paperclip. **Done:** already removed the Paperclip code
* Add support for Rails 6.0. **Done:** Changed the dependencies so we use the [ros-apartment gem](https://github.com/rails-on-services/apartment) instead of the unmaintained apartment gem
* Make sure sending emails plays nicely with multitenancy. **Done:** fixed the scenario with emails from different tenants sent simultaneously 
* Allow different ways to define valid postal codes for different tenants. **Done:** changed the code to use the changes in #2125
* Run `consul:execute_release_tasks` on every tenant. **Done:** rake tasks which should run on every tenant have been changed to do so
* Execute the processes defined in the `schedule.rb` file for each tenant. **Done:** the rake tasks have been updated (see previous point) and the sitemap generates files for each tenant
* Find a way to route `sitemap.xml` to the sitemap of the right tenant. **Done:** Make the `robots.txt` file link to a sitemap for the current tenant
* Right now,  tenants can be managed by administrators of any tenant. **Done:** changed so only administrators of the main tenant can manage them
* Check the cache is getting data from the right tenant. **Done:** use different cache namespaces per tenant
* Allow using different SMTP servers to send emails from different tenants. **Done:** use the secrets.yml file for per-tenant secrets
* Allow using different SMS settings for each tenant. **Done:** use the secrets.yml file for per-tenant secrets
* Allow different ManagerAuthenticator settings per tenant. **Done:** use the secrets.yml file for per-tenant secrets
* Configure authentication with social networks (including google and wordpress) for different tenants while keeping compatibility with our current way to do so. **Done:** use the secrets.yml file for per-tenant secrets
* Fix crash when creating/altering database extensions without superuser privileges. **Done:** we show a warning in this case
* Show a warning when the database user cannot create schemas. **Done:** we show it alongside the warnings about managing extensions
* Remove "list of tenants is empty" warning when working with single-tenant applications. **Done:** we only show it when multitenancy is enabled
* Make multitenancy work in applications running on subdomains (like `participate.mydomain.org`). **Done:** we now find the tenant for a request taking the default host (including subdomains) into account
* Make development seed data work with multitenancy. **Done:** we remove the tenant schemas before populating the default one with fresh data
* Update the installer so new installations install database extensions on the `shared_extensions` schema. **Done** in consul/installer#206 (pending merging)
* Allow running certain rake tasks on tenants other than the default one. **Done:** the `db:dev_seed` and budget tasks accept a parameter containing the name of a tenant
* Lower the risk of accidental database corruption if using a wrong `schema.rb` file. **Done:** in #5009
* Separate the attachments for different tenants. **Done:** Now attachments for different tenants are stored in different folders
* Make Machine Learning compatible with multitenancy. **Done:** Now data for different tenants is stored in different folders
* Add documentation regarding how to enable multitenancy and add notes about SSL certificates, domains, schema.rb, ... **Done** in consul/docs#116 (pending merging)
* Allow adding different image files in `app/assets/images/custom/` per tenant. **Partially done:** possible when using `image_path_for` to get the path of the image
* Fix wrong redirections with Devise controllers. **Done:** we now insert the Apartment middleware before the Warden middleware

### Future improvements

The following items are out of the scope of this pull request but will be implemented in different pull requests:

* Allow using diferent colors, header, footer, ... for each tenant: this will be done in pull request #5021; for now, we've simply added a different class to the `<html>` element depending on the current tenant
* Allow different tenants to have different domains and not just different subdomains (pull request #5007)
* Allow using different ERB files per tenant  (pull request #5030)
* Add a way to "disable" a tenant so it can no longer be accessed
* Allow enabling different languages for different tenants
* Right now, whenever a new tenant is created, a new admin with `admin@consul.dev` as email and `12345678` as password is created. Might be confusing and even a security risk. Maybe we could ask for an admin email when creating a tenant and (to simplify) generate a random password, so the first admin needs to reset their password

## Visual Changes

There's a new admin section listing existing tenants, with their name and subdomain

Tenants can be added and edited

## Release Notes

We've added support for multitenancy, that is, managing different institutions using just one CONSUL installation. Although we've done several tests in order to experience a smooth upgrade from previous version of CONSUL so this feature doesn't affect existing installations, as usual **we still recommend backing up yout database before upgrading**.

When upgrading your **development** environment, update your `config/database.yml` file and add the line `schema_search_path: "public,shared_extensions"` (you can copy it from the `config/database.yml.example` file). This is done automatically on production environments when running the `consul:execute_release_tasks` rake task.

If you'd like to enable multitenancy, **TODO:** add a link to the multitenancy documentation.

In the extremely unlikely case that, for some unusual reason, you've changed the default postgresql schema (which is named "public" by default) to something else, you'll have to change the code in the Apartment initializer and the Tenant model accordingly. If you don't know what this is about, you can safely ignore this paragraph :wink:. NOTE: this case is so unlikely that we might not even include it in the release notes.

### CODE REVIEW

File: app/models/tenant.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  def self.current
+    find_by(subdomain: Apartment::Tenant.current)
+  end
+
+  private
+
+    def create_tenant
+      unless subdomain == "public"
+        unless Rails.env.test?
+          Apartment::Tenant.create(subdomain)
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/models/tenant.rb:17]: False positive. This isn't an ActiveRecord object.

**Comment by robgitnyc** [comment]
Hi guys. Question. If I merge this branch as-is right now into a main branch and proceed to install, will I get a working multitenant instance of Consul?

**Comment by javierm** [comment]
Hi, @robgitnyc :smile:.

The branch is still in development; I've just updated the notes about the current status in the pull request description.

I think it's OK to do what you suggest on a staging environment in order to try this feature out. On production, use it at your own risk :wink:.

File: app/models/tenant.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  def self.current
+    find_by(subdomain: Apartment::Tenant.current)
+  end
+
+  private
+
+    def create_tenant
+      unless subdomain == "public"
+        Apartment::Tenant.create(subdomain)
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/models/tenant.rb:31]: False positive :wink:.

File: config/sitemap.rb [hunk 3]
```diff
@@ -6,49 +6,55 @@ def gzip(stream, data)
...
+    SitemapGenerator::Sitemap.sitemaps_path = nil
+  else
+    SitemapGenerator::Sitemap.sitemaps_path = "sitemaps/#{Tenant.current_subdomain}"
   end
 
-  if Setting["process.proposals"]
-    add proposals_path, priority: 0.7, changefreq: "daily"
-    Proposal.find_each do |proposal|
-      add proposal_path(proposal), lastmod: proposal.updated_at
+  SitemapGenerator::Sitemap.create do
```
[the remark this answers is not in the document]
Comment (javierm) [reply config/sitemap.rb:20]: False positive :wink:.

**Comment by DemocraciaPopular** [comment]
Hello guys. I tried to run this on local dev environment without success. The trace was long so I don't even know what to quote, lol.

Anyway, any time frame for this feature to be in master (and/or maybe some instructions to be able to run this locally meanwhile)?

**Comment by javierm** [comment]
Hi, @DemocraciaPopular :smile:.

> Hello guys. I tried to run this on local dev environment without success. The trace was long so I don't even know what to quote, lol.

Feel free to attach a file containing the whole backtrace (as long as it doesn't contain sensitive information) and we'll have a look.

> Anyway, any time frame for this feature to be in master (and/or maybe some instructions to be able to run this locally meanwhile)?

There are two different concepts in this pull request: production-ready for new installations and merged on the `master` branch. The difference is that before merging it in the master branch we need to make sure existing installations will keep working.

So the current schedule it to merge in the master branch during the first half of 2023. If you're doing a new CONSUL installation, you might use it much earlier, although of course you'll need to keep updating your code with the changes we do on this branch while it's in development.

Does it make sense? :wink:

**Comment by DemocraciaPopular** [comment]
> Hi, @DemocraciaPopular smile.
> 
> > Hello guys. I tried to run this on local dev environment without success. The trace was long so I don't even know what to quote, lol.
> 
> Feel free to attach a file containing the whole backtrace (as long as it doesn't contain sensitive information) and we'll have a look.
> 
> > Anyway, any time frame for this feature to be in master (and/or maybe some instructions to be able to run this locally meanwhile)?
> 
> There are two different concepts in this pull request: production-ready for new installations and merged into the `master` branch. The difference is that before merging it into the master branch we need to make sure existing installations will keep working.
> 
> So the current schedule it to have this feature in the master branch during the first half of 2023. If you're doing a new CONSUL installation, you might use it much earlier, although of course you'll need to keep updating your code with the changes we do on this branch while it's in development.
> 
> Does it make sense? wink

Hi. Yes, it start's making sense.

After some clarification of my local git branching, I could install the multitenant consul in local dev smoothly. It works wonderfully. I tried it with lvh.me and subdomains, and it's beautiful.

However, what doesn't work for me is the remote deploy. I try to deploy it to my production server and I get this error:
```
00:16 git:create_release
      01 mkdir -p /home/deploy/consul/releases/20220930221347
    ✔ 01 deploy@185.112.144.81 0.408s
      02 git archive working-multitenant-30-sept-2022 | /usr/bin/env tar -x -f - -C /home/deploy/consul/releases/20220930221347
      02 fatal: not a valid object name: working-multitenant-30-sept-2022
      02
      02 tar:
      02 This does not look like a tar archive
      02
      02 tar:
      02 Exiting with failure status due to previous errors
```

Now, I used to get this very error while trying to deploy master or stable, and it was an issue of not having updated the repo_url variable with the url of my fork in config/deploy.rb. Once that was corrected, error disapeared and deploy started working. However, now repo_url has the right url of my fork, but deploying multitenancy branch is failing. What could it be?

**Comment by DemocraciaPopular** [comment]
ok, my previous comment was getting messy after too many updates so I will sumarize here.

-multitenancy branch that has the "multitenancy" tab at the bottom (that's a commit from yesterday if i recall correctly) pushed to production works pretty decently in remote.
-latest commit as of right now doesn't work on remote, even though I could see the change in the interface that was moved to under the "settings" tab. I could see that when I ran it locally. 
-subdomains work in both remote configurations, even if in the second case you can't see the interface (they were created under the previously mentioned version).
-you can login with the main admin account to every tenant.
-registration emails from tenants are not sent, which means they don't share same smtp configuration, which means tenant's smtp details are stored somewhere else that I don't know.

Now the bad things:

During the day I was able to run every branch locally without issues, but after a while, too much experimentation maybe, now every time I run bin/rails from any branch i get a spec test ran instead. Those are extremely long and I have never seen the end of one.

That's what I can report so far.

**Comment by javierm** [comment]
@DemocraciaPopular I've updated the "Release Notes" section of this pull request with instructions on how to enable multitenancy. Please note this branch is currently under heavy development and we might break things at any time.

> I can login with the main admin account to every tenant.

Every time you create a tenant, the default admin account is created for that tenant. But you can change the password on one tenant (or delete the account on one tenant once you've created a different administrator) and it won't affect the other tenants.

> registration emails from tenants are not sent, which means they don't share same smtp configuration

Thanks for the feedback. We haven't tested this feature on production environtments yet (only on development); we'll do so during the next few weeks.

> i don't seem to find a difference in running either commit with schema_search_path: "public,shared_extensions" present or not in database.yml. Maybe I am missing something regarding to this file?

This change is needed when running tests dealing with the search feature and it might be needed when running certain rake tasks.

> now every time I run bin/rails from any branch i get a spec test ran instead.

You need to run `bin/rails s` (the `s` stands for `server`) :wink:.

**Comment by javierm** [comment]
@DemocraciaPopular We've done some testing regarding emails and tenants on staging, and the emails are being sent correctly. Maybe in your case the emails arrived in the spam folder when sending from the tnants, or maybe we've fixed whatever didn't work since the time you tried :thinking:.

**Comment by javierm** [comment]
@DemocraciaPopular We've investigated the issue with the emails a bit more. Make sure to use different `FROM` addresses for each tenant. You can configure these options in the "Settings" section of the administration:

Forms to fill in sender email name and address in the settings section

It looks like not doing so sometimes results in emails being identified as dangerous by some email providers.

File: spec/system/multitenancy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    with_subdomain("venus") do
+      visit polls_path
+
+      expect(page).to have_content "There are no open votings"
+      expect(page).to have_css "html.tenant-venus"
+      expect(page).not_to have_css "html.tenant-mars"
+    end
+  end
+
+  scenario "PostgreSQL extensions work for tenats" do
```
Comment (Senen) [inline spec/system/multitenancy_spec.rb:44]: Hi @javierm, typo found here:
```suggestion
  scenario "PostgreSQL extensions work for tenants" do
```

**Review by taitus — APPROVED** [review]

**Review by taitus — APPROVED** [review]

**Review by taitus — APPROVED** [review]

**Pull request merged**